### PR TITLE
Infrastructure as Code. Terraform

### PR DESCRIPTION
**Pull request by Durf1985**
## In the process done:

* Install Terraform
* Create provider
* Create VM instance
* Creating Firewall rules
* Creating global metadata key for all instances
* Planing and apply terraform template
* Creating output variables file
* Using `reddit-base` image with provisioners, instead `reddit-full` image
* Input vars file
* Create load balancer with parametrization

## How to launch a project:

```
terraform init
terraform plan
terraform apply
```

## How to check the performance:

```bash
gcloud compute addresses describe lb-ipv4-1 \                                                                                 
   --format="get(address)" \
   --global
```
Paste the received IP address into the address bar of your browser. You have to get through the load balancer into your reddit application

## PR checklist

The label with the homework number, in label description - homework lecture and topic

### CODE REVIEW

**Comment by Durf1985** [comment]
@loktionovam created VM instance via terraform
photo_2022-09-22_16-20-37
photo_2022-09-22_16-20-52

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,8 @@
 Question.txt
 TODO.txt
 variables.json
+*.tfstate
+*.tfstate.*.backup
+*.tfstate.backup
+*.tfvars
+.terraform/
```
Comment (loktionovam) [inline .gitignore:9]: add `.terraform.lock.hcl` also

File: terraform/.terraform.lock.hcl [hunk 2]
```diff
@@ -0,0 +1,21 @@
+# This file is maintained automatically by "terraform init".
```
Comment (loktionovam) [inline terraform/.terraform.lock.hcl]: remove this file because it auto generated by `terraform init` command and is not a part of your source code

**Review by loktionovam — CHANGES_REQUESTED** [review]

File: terraform/outputs.tf [hunk 3]
```diff
@@ -0,0 +1,3 @@
+output "app_external_ip" {
+  value = google_compute_instance.app.network_interface.0.access_config.0.nat_ip
```
Comment (loktionovam) [inline terraform/outputs.tf]: Add the description of this output

https://cloud.google.com/docs/terraform/best-practices-for-terraform#outputs

File: terraform/main.tf [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+  }
+  network_interface {
+    network = "default"
+    access_config {
+
+    }
+
+  }
+  metadata = {
+    ssh-keys = "appuser:${file("~/github_repo/devops_study_infra/packer/scripts/TXT.pub")}"
```
Comment (loktionovam) [inline terraform/main.tf]: This absolute path can be different on different computer (for example my path will be `~/src/github/bla-bla`) so replace it by relative path

File: terraform/lb.tf [hunk 5]
```diff
@@ -0,0 +1,116 @@
+resource "google_compute_instance_template" "default" {
+  name = "reddit-template"
+   disk {
+    auto_delete  = true
+    boot         = true
+    device_name  = "persistent-disk-0"
+    mode         = "READ_WRITE"
+    source_image = "projects/clgcporg2-118/global/images/reddit-full-1664802177"
```
Comment (loktionovam) [inline terraform/lb.tf]: here you should use your project variable rather than hardcoding

File: terraform/lb.tf [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+  machine_type = "e2-medium"
+  metadata = {
+    ssh-keys = "appuser:${file("${var.public_key_path}")}"
+  }
+  network_interface {
+    access_config {
+      network_tier = "PREMIUM"
+    }
+    network    = "global/networks/default" 
+    subnetwork = "regions/us-central1/subnetworks/default"
```
Comment (loktionovam) [inline terraform/lb.tf]: use region variable there rather than hardcoding

File: terraform/lb.tf [hunk 7]
```diff
@@ -0,0 +1,116 @@
+resource "google_compute_instance_template" "default" {
```
Comment (loktionovam) [inline terraform/lb.tf]: replace `default` resource name by something more suitable

File: terraform/lb.tf [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+  region = "us-central1-a"
+  scheduling {
+    automatic_restart   = true
+    on_host_maintenance = "MIGRATE" 
+    
+  }
+  tags = ["reddit-app"]
+}
+
+resource "google_compute_instance_group_manager" "default" {
```
Comment (loktionovam) [inline terraform/lb.tf]: replace `default` resource name by something more suitable

File: terraform/lb.tf [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+  version {
+    instance_template = google_compute_instance_template.default.id
+    name              = "primary"
+ 
+  }
+  base_instance_name = "vm"
+  target_size        = 2
+}
+
+resource "google_compute_firewall" "default" {
```
Comment (loktionovam) [inline terraform/lb.tf]: replace `default` resource name by something more suitable

File: terraform/lb.tf [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+  priority      = 1000
+  source_ranges = ["0.0.0.0/0"]
+  target_tags   = ["reddit-app"]
+  allow {
+    ports    = ["80"]
+    protocol = "tcp"
+  }
+}
+
+resource "google_compute_global_address" "default" {
```
Comment (loktionovam) [inline terraform/lb.tf]: replace `default` resource name by something more suitable

File: terraform/lb.tf [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+  name       = "lb-ipv4-1"
+  ip_version = "IPV4"
+}
+
+resource "google_compute_health_check" "default" {
```
Comment (loktionovam) [inline terraform/lb.tf]: replace `default` resource name by something more suitable

File: terraform/lb.tf [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+    port               = 9292
+    port_specification = "USE_FIXED_PORT"
+    proxy_header       = "NONE"
+    request_path       = "/"
+  }
+  timeout_sec         = 5
+  unhealthy_threshold = 2
+}
+
+resource "google_compute_backend_service" "default" {
```
Comment (loktionovam) [inline terraform/lb.tf]: replace `default` resource name by something more suitable

File: terraform/lb.tf [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+  timeout_sec                     = 30
+  backend {
+    group           = google_compute_instance_group_manager.default.instance_group
+    balancing_mode  = "UTILIZATION"
+    capacity_scaler = 0.7
+  }
+}
+
+
+resource "google_compute_url_map" "default" {
```
Comment (loktionovam) [inline terraform/lb.tf]: replace `default` resource name by something more suitable

File: terraform/lb.tf [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+  name            = "web-map-http"
+  default_service = google_compute_backend_service.default.id
+}
+
+resource "google_compute_target_http_proxy" "default" {
```
Comment (loktionovam) [inline terraform/lb.tf]: replace `default` resource name by something more suitable

File: terraform/lb.tf [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+  name    = "http-lb-proxy"
+  url_map = google_compute_url_map.default.id
+}
+resource "google_compute_global_forwarding_rule" "default" {
```
Comment (loktionovam) [inline terraform/lb.tf]: replace `default` resource name by something more suitable

**Review by loktionovam — CHANGES_REQUESTED** [review]

**Review by loktionovam — CHANGES_REQUESTED** [review]
You need to add `terraform/terraform.tfvars.example` because your tests are broken
https://github.com/Durf1985/devops_study_infra/commit/2c26adedd81d4403165630596bf54f05f6b2315d/checks

**Review by loktionovam — APPROVED** [review]

**Pull request merged**